twoliter/src/lock.rs

```diff
@@ -92,21 +119,186 @@ struct ManifestListView {
+    architecture: DockerArchitecture,
+}
+
+#[derive(Deserialize, Debug)]
+struct IndexView {
+    manifests: Vec<ManifestView>,
+}
+
+#[derive(Deserialize, Debug)]
+struct ManifestLayoutView {
+    layers: Vec<Layer>,
+}
+
+#[derive(Deserialize, Debug)]
+struct Layer {
+    digest: ContainerDigest,
+}
+
+#[derive(Debug)]
+struct ContainerDigest(String);
+
+impl<'de> Deserialize<'de> for ContainerDigest {
+    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let digest = String::deserialize(deserializer)?;
+        if !digest.starts_with("sha256:") {
+            return Err(D::Error::custom(format!(
+                "invalid digest detected in layer: {}",
+                digest
+            )));
+        };
+        Ok(Self(digest))
+    }
+}
+
+impl Display for ContainerDigest {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.write_str(self.0.as_str())
+    }
+}
+
+#[derive(Serialize, Debug)]
+struct ExternalKitMetadata {
+    sdk: LockedImage,
+    #[serde(rename = "kit")]
+    kits: Vec<LockedImage>,
+}
+
+#[derive(Debug)]
+struct OCIArchive {
+    image: LockedImage,
+    digest: String,
+    cache_dir: PathBuf,
+}
+
+impl OCIArchive {
+    fn new<P>(image: &LockedImage, digest: &str, cache_dir: P) -> Self
+    where
+        P: AsRef<Path>,
+    {
+        Self {
+            image: image.clone(),
+            digest: digest.into(),
+            cache_dir: cache_dir.as_ref().to_path_buf(),
+        }
+    }
+
+    fn archive_path(&self) -> PathBuf {
+        self.cache_dir.join(self.digest.replace(':', "-"))
+    }
+
+    async fn pull_image(&self) -> Result<()> {
+        let digest_uri = self.image.digest_uri(self.digest.as_str());
+        let oci_archive_path = self.archive_path();
+        if !oci_archive_path.exists() {
+            let oci_archive_str = oci_archive_path.to_string_lossy();
+            docker!(
+                ["save", digest_uri.as_str(), "-o", oci_archive_str.as_ref()],
+                format!(
+                    "failed to fetch kit and save to disk from {} to {}",
+                    digest_uri, oci_archive_str
+                )
+            );
+        }
+        Ok(())
+    }
+
+    async fn unpack_layers<P>(&self, out_dir: P) -> Result<()>
+    where
+        P: AsRef<Path>,
+    {
+        let path = out_dir.as_ref();
+        let digest_file = path.join("digest");
+        ensure!(digest_file.exists(), "digest file '{}' does not exist, most likely because the oci archive has not been pulled", digest_file.display());
+        let digest = read_to_string(&digest_file).await.context(format!(
+            "failed to read digest file at {}",
+            digest_file.display()
+        ))?;
+        if digest == self.digest {
+            return Ok(());
+        }
+
+        remove_dir_all(path).await?;
+        create_dir_all(path).await?;
+        write(&digest_file, self.digest.as_str())
```

Comment:
Now this should be the last operation, since if this write succeeds but a later step fails, then we'll never retry because the digest will match.